### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Lada496**
#### What I learned
- [learned memorization](https://github.com/Lada496/stand-on-the-sholders-of-giants/commit/b29e6ea9661f9cf893db8997ab2be8d5ec8323b2): a technique to cache data
- [learned currying](https://github.com/Lada496/stand-on-the-sholders-of-giants/commit/31a74726c9bf31ababcb2824bb56683c5d647d79):  the technique of translating the evaluation of a function that takes multiple arguments
- [learned types of tests and testing libraries](https://github.com/Lada496/stand-on-the-sholders-of-giants/commit/35c09da530956de46a9ab3c69b78c8f0c8c953c4)
- set up pre-commit with husky
- [learned serverless programming with stripe and netlify](https://github.com/Lada496/stand-on-the-sholders-of-giants/commit/94a8a5a21ad33ada2d154f70c3899dcc8b6635f4)

#### What I thought
- I was stuck to set up husky because this repo includes tests for React components. I temporary decided test only in test-in-react folder for now but I'm sure I have to set up something to run test with react component as well (I guess it should be babel and webpack). Let me know if you have any suggestion. 

#### Others
- Thank you for providing code examples for  [type inheritance techniques](https://www.typescriptlang.org/play?#code/PTAEF5K7tAzA9ggdAFwM4RtyAoAlgHaoCmATnAIYDGJoAYkgEonoAOCh6JAQggCYBPUAG9coUPnSMEALlAAjJABsSlQgG5x8JABV8qVfPSoyRAOZaJ+fsdMWtAXy25UgtnRYBHAK6tUAMqolKQQoABETACiAIoAqlEBugCSAHIA4uGgAD4RACJR9ACCcQAyulm54cXJpVF54S5EpBQ0niS+-nxCADy64D6EANaECADuhAB8otpKQvK6VqBkHX4mQSEk8t5rgcGkTri4IFg4MKCoSACyPtQAFsmEd+QG6rRo6AAep2dQRye6O5SSTEchUWigO6UTBEZ5mUj8UCUZTKC7PUBjSjCOBkBAAWwYzFYHC4vAEwnA0x2XXJEGmAXxJAAwviSSRiOgAAq4tjoY5gZKoADkmFGqCRhFAJFU5nU4vQCGUPlQ+E4SMwqCBMMwQyIiIQcAi5hI4oU6kI6koimVUsI-BIiJ8bFAbB8KIsaLoiBWJiy6AMPhCqsIBFBrQhumutweTxewUItG5CF5Us+pDtmFA1JM3UEPRkLHYnG4uemYgkKsMW1AJjMhEs2k43NY6HkAAoAJR00AANwQNkO-OWamo4uorM47PFAFpQIIED5QIQSA6XZQyCrkRd3KwADTW8XzxfrugmfAo0DG1Aq+tIi+wuOrtwePlwQaj4OgSMIG73R5w14ExINsRErVR9ybH10H3FZOnWfYSH3OZBHkEQbEcRwFijP9Y3hN4SCTXkuzEbQThWVAfDILga0ZKVVDxKdtHIyjNFwRx-jAX5zlUdB0H-ON8I+b4uOgDivy1EEWnBOg2HXTcUWEB94UwBUGO3F8D0kcV+AQbhCCFcUoR7Og8TdFU2FUNcVg5IdBRFGtlHwcw7kMYQFBNFp1VAABGDsYLUfhKAUSyACY-NTCz8GoAxQAAZnCnTWH0sdOGCIhQEGZdaF49dhDYHlyBVVhkFDKS2lAUoW34vCgMIzBywuAwjBrex6yWSCW3bLtKV7ft+CWGx5ALYlizJIQAG1whscIAF0llg3YNlIbZVn8JaSEHMiRxSvE2WIUBZ3y5NMBPRQYrxBAVlAahVBPMgJURC0yFxMZcDfBMVTVAISHHO0WV2ydiBAsDENADreP3Gx-LgvZNkwiqqtwwDEx5dBiKOCQtooqiVNo6USAY4gmJNFinCAA): <br />
I saw now techs here are questions.
1️⃣ `type RequestState = "REQUESTING" | "DEFAULT" | "FAILED";` Is this an alternative for enum? <br />
2️⃣ Can we pick a type inside of a type like this `id: FooResponseBody["id"];`?
- I'm 100% positive to invite a new person if he wants. (I assume he has more experience than I have, so I'm not sure is there anything he can learn from this though). I'll ask him if we can see projects each other.

### CODE REVIEW

**Comment by Lada496** [comment]
typo 🤦‍♀️
not "now techs" but "new techs"

File: package.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "stand-on-the-sholders-of-giants",
+  "version": "1.0.0",
+  "description": "This is a root repository for the mentorship",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
```
Comment (kei95) [inline package.json:8]: Could it be the reason why your husky breaks during pre-commit hooks?

**Comment by kei95** [comment]
> I was stuck to set up husky because this repo includes tests for React components. I temporary decided test only in test-in-react folder for now but I'm sure I have to set up something to run test with react component as well (I guess it should be babel and webpack). Let me know if you have any suggestion.

I'm not sure why husky has something to do with `jest`. It's hard to say what it is without error message or stack trace.

> 1️⃣ type RequestState = "REQUESTING" | "DEFAULT" | "FAILED"; Is this an alternative for enum?

Sort of. This is just a type that specifies certain strings. Like it could've been `RequestState = string` but obviously, it's too general because even a million characters could be `string` at the end of the day. So if you want to expect a certain type of `string`, it's better to declare its exact shape of it.

> 2️⃣ Can we pick a type inside of a type like this id: FooResponseBody["id"];?

For sure! This is how to access and grab an object-shaped type's values. This is how you do;
```typescript
interface Foo {
  title: "TITLE";
  id: number | null
}

type FooTytle = Foo["title"]; // "TITLE"
```

> I'm 100% positive to invite a new person if he wants. (I assume he has more experience than I have, so I'm not sure is there anything he can learn from this though). I'll ask him if we can see projects each other.

Great thanks! I think you both are on a similar level so it'd be okay. He should be able to learn quite a few stuff from your learning materials.

**Comment by kei95** [comment]
Good job as always! Getting a sense of what the effective approach for testing takes some time to develop, but it certainly makes you a great developer if you can consider it when you write code👍

**Pull request merged**